From 6168ec6e5353163c29450bf4c805d6c79c5689e0 Mon Sep 17 00:00:00 2001
From: pmandri <padmashree.mandri@intel.com>
Date: Mon, 19 Aug 2024 16:14:20 +0530
Subject: [PATCH] Dynamic detection of primary sound card

This patch adds changes to detect primary sound card dynamically.
And hence avoids usb cards getting detected as primary card, when
they are listed as 0th card.

Tests Done:
USB playback and Record
Camera record

Tracked-On: OAM-123512
Signed-off-by: pmandri <padmashree.mandri@intel.com>
---
 audio/aidl/default/primary/PrimaryMixer.h    | 14 ++++++++++++++
 audio/aidl/default/primary/StreamPrimary.cpp |  8 ++++++--
 2 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/audio/aidl/default/primary/PrimaryMixer.h b/audio/aidl/default/primary/PrimaryMixer.h
index 3806428cfd..79f3fa844a 100644
--- a/audio/aidl/default/primary/PrimaryMixer.h
+++ b/audio/aidl/default/primary/PrimaryMixer.h
@@ -34,6 +34,20 @@ class PrimaryMixer : public alsa::Mixer {
     static constexpr int kAlsaDevice = 0;
 
     static PrimaryMixer& getInstance();
+    static int get_pcm_card(const char* name)
+    {
+        char id_filepath[20] = {0};
+        char number_filepath[20] = {0};
+        ssize_t written;
+
+        snprintf(id_filepath, sizeof(id_filepath), "/proc/asound/%s", name);
+
+        written = readlink(id_filepath, number_filepath, sizeof(number_filepath));
+        if (written < 0)
+                return -1;
+
+        return atoi(number_filepath + 4);
+    }
 
   private:
     PrimaryMixer() : alsa::Mixer(kAlsaCard) {}
diff --git a/audio/aidl/default/primary/StreamPrimary.cpp b/audio/aidl/default/primary/StreamPrimary.cpp
index 7325a91b49..2b673604f2 100644
--- a/audio/aidl/default/primary/StreamPrimary.cpp
+++ b/audio/aidl/default/primary/StreamPrimary.cpp
@@ -20,6 +20,7 @@
 #include <audio_utils/clock.h>
 #include <error/Result.h>
 #include <error/expected_utils.h>
+#include <unistd.h>
 
 #include "PrimaryMixer.h"
 #include "core-impl/StreamPrimary.h"
@@ -92,13 +93,16 @@ StreamPrimary::StreamPrimary(StreamContext* context, const Metadata& metadata)
 }
 
 std::vector<alsa::DeviceProfile> StreamPrimary::getDeviceProfiles() {
+    int primary_card_num  = primary::PrimaryMixer::get_pcm_card("PCH");
+    LOG(INFO) << __func__ << "Primary Card number is:" << primary_card_num;
+
     static const std::vector<alsa::DeviceProfile> kBuiltInSource{
-            alsa::DeviceProfile{.card = primary::PrimaryMixer::kAlsaCard,
+            alsa::DeviceProfile{.card = primary_card_num,
                                 .device = primary::PrimaryMixer::kAlsaDevice,
                                 .direction = PCM_IN,
                                 .isExternal = false}};
     static const std::vector<alsa::DeviceProfile> kBuiltInSink{
-            alsa::DeviceProfile{.card = primary::PrimaryMixer::kAlsaCard,
+            alsa::DeviceProfile{.card = primary_card_num,
                                 .device = primary::PrimaryMixer::kAlsaDevice,
                                 .direction = PCM_OUT,
                                 .isExternal = false}};
-- 
2.34.1

